calyx/src/passes/static_par_conv.rs

```diff
@@ -0,0 +1,252 @@
+/// apply this transformation to a subset of seq blocks in the par block.
+/// Regardless, this transformation should never increase the number of cycles the par
+/// block takes to execute.
+///
+/// #Example
+/// par {
+///    seq { @static(M) A0; @static(N) B0; @static(P) C0; }
+///    seq { @static(M) A1; @static(N) B1; @static(P) C1; }
+///    seq { @static(M) A2; @static(N) B2; @static(P) C2; }
+///}
+///
+/// into
+///
+/// @static(M + N + P) seq {
+///     @static(M) par{ @static(M) A0; @static(M) A1; @static(M) A2;}
+///     @static(N) par{ @static(N) B0; @static(N) B1; @static(N) B2;}
+///     @static(P) par{ @static(P) C0; @static(P) C1; @static(P) C2;}
+/// }
+
+pub struct StaticParConv;
+
+impl Named for StaticParConv {
+    fn name() -> &'static str {
+        "static-par-conv"
+    }
+
+    fn description() -> &'static str {
+        "Transform `par` of `seq` to `seq` of `par` under correct conditions"
+    }
+}
+
+// given a stmt, returns Some(&val) where val is the value of the "static"
+// attribute of stmt. Returns None if no "static" attribute exists.
+fn get_static_attr(stmt: &ir::Control) -> Option<&u64> {
+    stmt.get_attributes().and_then(|atts| atts.get("static"))
+}
+
+// Takes two seqs, longer and shorter. longer should be at least as long as
+// shorter. Returns Some(vec) if there exists arrangement of shorter and longer
+// such that each statement in shorter can be paired with a statement in longer,
+// such that executing these pairs as a seq of pars will respect dependencies and
+// the number of cycles it takes will remain the same. If no such vec exists, returns None.
+// The vec it returns contains which indices of longer each element in shorter
+// should be paired with. So, if we represent the vecs by the static attribute of their
+// statements, and longer = [4,1,5] and shorter = [4,5] then vec would be
+// [0,2], since the 4 in shorter can be paired with the 4 in longer (@ index 0)
+// and the 5 in shorter can be paired with the 5 in longer (@ index 2).
+// A consequence of this is that vec should always be the same length as shorter.
+fn is_compatible(longer: &ir::Seq, shorter: &ir::Seq) -> Option<Vec<usize>> {
+    let mut long_iter = (*longer).stmts.iter();
+    let mut short_iter = (*shorter).stmts.iter();
+
+    let mut long_val = long_iter.next();
+    let mut short_val = short_iter.next();
+
+    let mut index_counter = Vec::new();
+    let mut counter = 0;
+
+    while let (Some(c1), Some(c2)) = (long_val, short_val) {
+        match (get_static_attr(c1), get_static_attr(c2)) {
+            (Some(x1), Some(x2)) => {
+                if x2 <= x1 {
+                    short_val = short_iter.next();
+                    index_counter.push(counter);
+                }
+                long_val = long_iter.next();
+            }
+            _ => return None,
+        }
+        counter += 1;
+    }
+
+    match short_val {
```

Comment:
You can use `Option::map` here: `short_val.map(|_| index_counter)`

Reply:
Ah, I just realized that `Option::map` maps Some -> Some and None-> None. 
But I want to only return Some(index_counter) when `short_val` is None,(since means that `short_val` has been fully iterated). So I don't think `Option::map` quite works here.